pkg/hub/lease/controller.go

```diff
@@ -0,0 +1,180 @@
+}
+
+// sync checks the lease of each accepted cluster on hub to determine the managed cluster whether is available.
+func (c *leaseController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	clusters, err := c.clusterLister.List(labels.Everything())
+	if err != nil {
+		return nil
+	}
+	for _, cluster := range clusters {
+		// cluster is not accepted, skip it.
+		acceptedCondition := helpers.FindManagedClusterCondition(cluster.Status.Conditions, clusterv1.ManagedClusterConditionHubAccepted)
+		if !helpers.IsConditionTrue(acceptedCondition) {
+			continue
+		}
+
+		leaseDurationSeconds := cluster.Spec.LeaseDurationSeconds
+		// TODO: use CRDs defaulting or mutating admission webhook to eliminate this code.
+		if leaseDurationSeconds == 0 {
+			leaseDurationSeconds = 60
+		}
+
+		// check the cluster lease whether is constantly updated in a grace period.
+		// the grace period will be 5 times cluster lease duration.
+		availableCondition, err := c.checkLease(cluster, time.Duration(leaseDurationTimes*leaseDurationSeconds)*time.Second)
+		if err != nil {
+			return err
+		}
+
+		availableConditionUpdateFn := helpers.UpdateManagedClusterConditionFn(*availableCondition)
+		_, updated, err := helpers.UpdateManagedClusterStatus(ctx, c.clusterClient, cluster.Name, availableConditionUpdateFn)
+		if err != nil {
+			return err
+		}
+		if updated {
+			syncCtx.Recorder().Eventf("ManagedClusterAvailableConditionUpdated", "update managed cluster %q available condition to %q, due to %q",
+				cluster.Name, availableCondition.Status, availableCondition.Reason)
+		}
+	}
+	return nil
+}
+
+// checkLease checks a given cluster's lease and determine the cluster whether is available.
+// 1. if the cluster does not have a lease, update the cluster available condition to unknown
+// 2. if the cluster have a lease and does not have an available condition, update the cluster
+//    available condition to unknown
+// 3. if the cluster have a lease and have an available condition
+//   - if the lease is constantly updated, keep the cluster available condition
+//   - if the lease is not constantly updated in a grace period, update the cluster available
+//     condition to unknown
+func (c *leaseController) checkLease(cluster *clusterv1.ManagedCluster, gracePeriod time.Duration) (*clusterv1.StatusCondition, error) {
+	clusterLeaseName := fmt.Sprintf("cluter-%s-lease", cluster.Name)
+	observedLease, err := c.leaseLister.Leases(cluster.Name).Get(clusterLeaseName)
+	if errors.IsNotFound(err) {
+		return &clusterv1.StatusCondition{
+			Type:    clusterv1.ManagedClusterConditionAvailable,
+			Status:  metav1.ConditionUnknown,
+			Reason:  "ManagedClusterLeaseNeverUpdated",
+			Message: "Registration agent never updated its lease.",
+		}, nil
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	observedCond := helpers.FindManagedClusterCondition(cluster.Status.Conditions, clusterv1.ManagedClusterConditionAvailable)
+	if observedCond == nil {
```

Comment:
I don't think you need to check this before continuing.  I think that the two flows are independent.

Reply:
Should the klusterlet actually be writing this field ever? Shouldn't it update the lease when it first comes online?

Reply:
remove this check, the klusterlet should update its lease when it first comes online